quickwit/quickwit-common/src/semaphore_with_waiter.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright (C) 2023 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+
+use tokio::sync::{OwnedSemaphorePermit, Semaphore, TryAcquireError};
+
+/// `SemaphoreWithLimitedWaiters` is an extension of semaphore
+/// that limits the number of waiters.
+///
+/// If more than n-waiters then acquire returns an error.
+#[derive(Clone)]
+pub struct SemaphoreWithLimitedWaiters {
+    permits: Arc<Semaphore>,
+    // Implementation detail:
+    // We do not use the async semaphore mechanics.
+    // The waiter count could have implemented using a simple atomic counter.
+    // We use a semaphore to avoid the need to reimplement the looping compare-and-swap dance.
+    waiter_count: Arc<Semaphore>,
+}
+
+impl SemaphoreWithLimitedWaiters {
+    /// Creates a new `SemaphoreWithLimitedWaiters`.
+    pub fn new(num_permits: usize, max_num_waiters: usize) -> Self {
+        Self {
+            permits: Arc::new(Semaphore::new(num_permits)),
+            waiter_count: Arc::new(Semaphore::new(max_num_waiters)),
+        }
+    }
+
+    /// Acquires a permit.
+    pub async fn acquire(&self) -> Result<OwnedSemaphorePermit, ()> {
+        match Semaphore::try_acquire_owned(self.permits.clone()) {
+            Ok(permit) => {
+                return Ok(permit);
+            }
+            Err(TryAcquireError::NoPermits) => {}
+            Err(TryAcquireError::Closed) => {
+                panic!("semaphore closed (this should never happen)");
```

Comment:
```suggestion
                // The `permits` semaphore should never be closed because we don't expose the `close` API and never call `close` internally.
                panic!("semaphore should not be closed");
```